Comprehensive Integration Test Suite

RyanL2004 opened the pull request:
_Please read #139_ 

### Overview 
This PR  introduces a robust integration test suite for GitIngest, covering critical functionalities and edge cases. These tests ensure the reliability, scalability, and correctness of the application.

### What Does This PR Do?
#### 1. Adds Comprehensive Tests
- Tests for analyzing remote repositories (e.g., input validation, processing, and output generation).
- Tests for handling edge cases, including invalid URLs, special characters, and unsupported inputs.
- Tests for analyzing large repositories and handling nested structures.

#### 2. Simulates Concurrent Requests
- Validates the server’s ability to handle multiple simultaneous requests without failures.
- Tests rate limiting and stability under load.

#### 3. Mocks Template Rendering
- Bypasses template file dependencies, ensuring faster and isolated testing of core logic.

#### 4. Improves Code Coverage
- Ensures that all major code paths, including edge cases and error scenarios, are tested.

#### 5. Ensures Graceful Error Handling
- Validates proper user feedback for invalid inputs and large files.

### How Does It Work?
- **Test Fixtures**: Mocked templates and a test client simulate real-world scenarios while ensuring fast and isolated tests.
- **Concurrent Requests**: Thread pooling simulates high traffic, validating server stability.
- **Edge Case Tests**: Specific inputs test the system’s robustness against unusual and unsupported cases.

### What Does It Test?
#### 1. End-to-End Flows
- Repository analysis from form submission to result rendering.

#### 2. Error Handling
- Invalid inputs, large files, and unsupported patterns.

#### 3. Performance
- Large repositories and high concurrency.

#### 4. Scalability
- Stability under simultaneous requests.

### Why Is This Critical?
1. Reduces the risk of bugs and regressions.
2. Enhances developer confidence in making changes.
3. Improves user experience by validating common and edge-case scenarios.
4. Ensures operational stability under load.

### Conclusion
This PR significantly enhances GitIngest’s testing framework, ensuring robustness, reliability, and scalability. It provides a foundation for future development with confidence in the system’s core functionalities.

RyanL2004 commented:
I wanted to provide some context regarding the previous failures on this PR:

1. **Initial Failure**:
   The CI tests initially failed because the application was attempting to mount a `static` directory using this line in `main.py`:
   ```python
   app.mount("/static", StaticFiles(directory="static"), name="static")```
However, the static directory was missing in the CI environment, resulting in the following error:

 `RuntimeError: Directory 'static' does not exist`
 
 2. **Fix**:
    I corrected the directory path to `src/static` since that’s the actual location of the`static` folder in the project. Here’s the updated line in `main.py`:
 ```python
 app.mount("/static", StaticFiles(directory="src/static"), name="static")
 ```
3. **Resolution**:
  After making this change, I pushed the fix to this branch, which resolved the CI errors. The tests now pass successfully, and the PR is ready for review.

cyclotruc reviewed (changes requested):
Hey @RyanL2004 thank you very much for this work

I'll take a deeper look at the tests themselves very soon but there's an issue that needs to be tackled first:
![image](https://github.com/user-attachments/assets/6887c78f-b429-4c75-ba35-fceedd4e04dd)

The changes you made in `main.py` are  breaking the  web server

That makes me think that this is the first thing we should try to test: 
- is the server runnable and is it possible to trigger an ingest?

Feel free to reach out on discord if you need any help in structuring those new tests

RyanL2004 commented:
Hi @cyclotruc,
Thank you for your feedback and for taking the time to review the work!

I've addressed the issue regarding the `main.py` changes and the web server breaking:

1. **Issue**:
   The web server failed to run due to an incorrect path for the `static` directory in the following line:
   ```python
   app.mount("/static", StaticFiles(directory="static"), name="static")
   ```
   The directory `static` was assumed to be at the root level, but it was actually located in `src/static`.

**Fix**: I’ve updated `main.py` to dynamically resolve the correct path for the `static` directory relative to `main.py`:

```python
from pathlib import Path

static_dir = Path(__file__).parent / "static"
app.mount("/static", StaticFiles(directory=static_dir), name="static")

cyclotruc commented on src/main.py:
While this solution might work as intended, the way it tackles the static file issue looks a lot like a hack and I'm not sure this is necessary
the fix should be on the test side in order to not modify the existing behaviour that we want to test

in other words: `main.py` should not change, unless we clearly identify a valid reason to change it

cyclotruc replied:
Maybe I'm missing a point here and feel free to explain to me the elements I might be missing to properly consider the changes

RyanL2004 replied:
Hi @cyclotruc ,
Indeed I agree, the if statement is made to tackle edge cases where the where the static file is not found or none existent meaning it indeed acts more as a hack rather than a fix, the real answer to that is there is no real fix to provide when path directory is not found.
That said it deleted it since the If statement didn't really matter as my tests would still run being able to find the static file accordingly and mounting the app using the if `/static` instead of the else case ` static file doesn't exist`.

I refactored the main.py to delete the if statement.

Now regarding this line i refactored:
```python
app.mount("/static", StaticFiles(directory=static_dir), name="static")
```

The pupose of it is to tell the FastAPI application to serve files from the `static` directory whenever users access URLs starting with /static.

Why is it needed:
By using `Path(__file__).parent / "static` instead of just `/static`, the application can adapt its location ensuring it can always find the static directory (ofc if its available somewhere in the project) regardless of where the code is deployed. 

You can try it urself and see the enhancements it offers.
As well as all the integration test cases running smoothly and accordingly

cyclotruc replied:
Thanks for the clarification!!
